src/ipc/ipc4/helper.c

```diff
@@ -371,28 +371,116 @@ int ipc_buffer_free(struct ipc *ipc, uint32_t buffer_id)
 int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 {
 	int dir = PPL_CONN_DIR_BUFFER_TO_COMP;
-	struct comp_dev *src, *dst;
-	struct comp_buffer *buffer;
+	struct ipc4_module_bind_unbind *bu;
+	struct comp_buffer *buffer = NULL;
+	struct comp_dev *src, *sink;
+	struct list_item *sink_list;
+	int src_id, sink_id;
 	uint32_t flags;
+	int ret;
+
+	bu = (struct ipc4_module_bind_unbind *)_connect;
+	src_id = IPC4_COMP_ID(bu->header.r.module_id, bu->header.r.instance_id);
+	sink_id = IPC4_COMP_ID(bu->data.r.dst_module_id, bu->data.r.dst_instance_id);
+	src = ipc4_get_comp_dev(src_id);
+	sink = ipc4_get_comp_dev(sink_id);
+	if (!src || !sink) {
+		tr_err(&ipc_tr, "failed to find src %x, or dst %x", src_id, sink_id);
+		return IPC4_INVALID_RESOURCE_ID;
+	}
 
-	src = ipc4_get_comp_dev(_connect[0]);
-	dst = ipc4_get_comp_dev(_connect[1]);
+	list_for_item(sink_list, &src->bsink_list) {
+		int src_queue = bu->data.r.src_queue;
+		struct comp_buffer *buf;
 
-	if (!src || !dst) {
-		tr_err(&ipc_tr, "failed to find src %x, or dst %x", _connect[0], _connect[1]);
+		buf = container_of(sink_list, struct comp_buffer, source_list);
+		if (buf->id == src_queue) {
+			buffer = buf;
+			break;
+		}
+	}
+
+	if (!buffer)
 		return IPC4_INVALID_RESOURCE_ID;
+
+	irq_local_disable(flags);
+	list_item_prepend(buffer_comp_list(buffer, dir), comp_buffer_list(sink, dir));
+	buffer_set_comp(buffer, sink, dir);
+	comp_writeback(sink);
+	irq_local_enable(flags);
+
+	dcache_writeback_invalidate_region(buffer, sizeof(*buffer));
+
+	ret = src->drv->ops.cmd(src, IPC4_MODULE_BIND, bu, sizeof(*bu));
```

Comment:
Why are we calling cmd() on the component here ? Isn't connecting the pipeline list enough ?

Reply:
the problem is module bind can't be fully match up with component connection. For sof, buffer is created by topology, but it is maintained by module in FW for IPC4 case, so sof component connect can be done by ipc message but module bind need the help of module

Reply:
ok, but rather than using cmd() callback can we use the same callback API that module interface uses here ?

Reply:
you mean to add bind_unbind interface in ops ?

Reply:
No I mean what is the API use for external modules to do this ? I cant see a bind_unbind API for external module ?

Reply:
All modules are based on parent common module  class which has ConnectOutPin and onBind, onUnBind functions. ConnectOutPin will be done in parent common module  class to set up in & out relation. And onBind & onUnBind will be dispatch to each sub module class.  external modules also inherit from this  common module class, so these are applied to it.

Reply:
how to create a external module
(1) create a module inheriting from common module class which will do common task for external module
(2) design and generate all .txt, .data. .bss segment in .X file
(3) create a config file for FW image generator such as rimage
(4) update Makefile
(5) implement LargeConfigSet (cmd), onPplState (like trigger), copy function.

Reply:
since it is not reasonable to  require external modules to create buffer by its own,  I decide to to maintain buffer by ipc framework not by modules.  buffer id will includes both input & output queue id for module to copy data to matched buffers.

And we still need to call cmd to support onBind & onUnBind event.

Reply:
> All modules are based on parent common module class which has ConnectOutPin and onBind, onUnBind functions.

ok, lets add C versions of these to comp_dev.

Reply:
so do we need to add onBind & unBind in drv->ops , or use ops->cmd with bind & unbind cmd ?

Reply:
lets add the bind and unbind.